File: Data Structures & Algorithms/shortest-path-in-binary-matrix/submission-2.kt
class Solution {
    fun shortestPathBinaryMatrix(grid: Array<IntArray>): Int {

        if (grid[0][0] == 1) {
            return -1
        }

        val directions = arrayOf(
            intArrayOf(0, 1),
            intArrayOf(1, 0),
            intArrayOf(0, -1),
            intArrayOf(-1, 0),
            intArrayOf(1, 1),
            intArrayOf(-1, -1),
            intArrayOf(1, -1),
            intArrayOf(-1, 1)
        )

        var stack = mutableListOf<IntArray>()
        var visited = mutableSetOf<String>()
        stack.add(intArrayOf(0, 0, 1))
        visited.add("${0}-${0}")

        while (stack.count() > 0) {
            val size = stack.count()
            for (i in 0 ..< size) {
                val cell = stack.removeFirst()
                if (cell[0] == grid.count() - 1 && cell[1] == grid[0].count() - 1) {
                    return cell[2]
                }
                for (direction in directions) {
                    val r = cell[0] + direction[0]
                    val c = cell[1] + direction[1]
                    if (r >= 0 && c >= 0 && r < grid.count() && c < grid[0].count() && !visited.contains("${r}-${c}") && grid[r][c] == 0) {
                        stack.add(intArrayOf(r, c, cell[2] + 1))
                        visited.add("${r}-${c}")
                    }
                }
            }
        }

        return -1

    }
}

File: Data Structures & Algorithms/swim-in-rising-water/submission-2.swift
class Solution {
    func swimInWater(_ grid: [[Int]]) -> Int {

        struct Node: Comparable {
            let x: Int
            let y: Int
            let time: Int

            static func < (n1: Node, n2: Node) -> Bool {
                return n1.time < n2.time
            }
        }
        
        var minHeap = Heap<Node>()

        var visited = Set<String>()
        var directions = [
            [0, 1],
            [0, -1],
            [1, 0],
            [-1, 0]
        ]

        minHeap.insert(Node(x: 0, y: 0, time: grid[0][0]))

        while minHeap.count > 0 {
            var node = minHeap.popMin()!
            var x = node.x
            var y = node.y
            var time = node.time
            if (visited.contains("\(x)-\(y)")) {
                continue
            }
            if x == grid.count - 1 && y == grid[0].count - 1 {
                return time
            }
            visited.insert("\(x)-\(y)")
            for direction in directions {
                var newX = direction[0] + x
                var newY = direction[1] + y
                if newX >= 0 && newX < grid.count && 
                newY >= 0 && newY < grid[0].count && !visited.contains("\(newX)-\(newY)") {
                    var nextElevation = grid[newX][newY]
                    minHeap.insert(Node(x: newX, y: newY, time: max(time, nextElevation)))
                }
            }
        }

        return -1

    }
}